[JSC] Fix wrong member access due to DECLARE_VISIT_CHILDREN macro

kljohann opened the pull request:
#### b0498801e751395d6776b6d459d2010737c217b1
<pre>
[JSC] Fix wrong member access due to DECLARE_VISIT_CHILDREN macro
<a href="https://bugs.webkit.org/show_bug.cgi?id=296688">https://bugs.webkit.org/show_bug.cgi?id=296688</a>

Reviewed by Yusuke Suzuki.

Using the DECLARE_VISIT_CHILDREN macro in private/protected sections can leak
subsequent members, as the macro ends with a public access specifier.

This change moves all uses of the macro into the public section of their
respective classes and selectively makes members used in existing
code public.

* Source/JavaScriptCore/runtime/JSBoundFunction.h:
Move DECLARE_VISIT_CHILDREN macro to the public section.
(JSC::JSBoundFunction::isTainted): New public getter for m_isTainted.
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::executeCall): Use new isTainted() getter.

* Source/JavaScriptCore/runtime/IntlDateTimeFormat.h:
Move DECLARE_VISIT_CHILDREN macro to the public section.
* Source/JavaScriptCore/runtime/IntlDateTimeFormatPrototype.cpp:
(JSC::handleDateTimeValue): Made TU-local as it is only used by host
functions within this file.
(JSC::IntlDateTimeFormat::handleDateTimeValue): Deleted.

* Source/JavaScriptCore/runtime/IntlNumberFormat.h:
Move DECLARE_VISIT_CHILDREN macro to the public section.
Add missing friend declarations.
(JSC::IntlNumberFormat::trailingZeroDisplayString): Made public for
use in IntlPluralRules::resolvedOptions.

* Source/JavaScriptCore/runtime/IntlPluralRules.h:
Move DECLARE_VISIT_CHILDREN macro to the public section.
Add missing friend declaration.

* Source/JavaScriptCore/runtime/RegExpObject.h:
Move DECLARE_VISIT_CHILDREN macro to the public section.
(JSC::RegExpObject::lastIndexIsWritable): Make public for
use in regExpProtoFuncSearch.

* Source/JavaScriptCore/tools/JSDollarVM.cpp:
(JSC::JSDollarVM::WasmStreamingCompiler): Add missing private
access specifier.

* Source/JavaScriptCore/API/JSCallbackObject.h:
* Source/JavaScriptCore/bytecode/UnlinkedCodeBlock.h:
* Source/JavaScriptCore/runtime/AbstractModuleRecord.h:
* Source/JavaScriptCore/runtime/EvalExecutable.h:
* Source/JavaScriptCore/runtime/GenericArgumentsImpl.h:
* Source/JavaScriptCore/runtime/IntlCollator.h:
* Source/JavaScriptCore/runtime/IntlLocale.h:
* Source/JavaScriptCore/runtime/IntlRelativeTimeFormat.h:
* Source/JavaScriptCore/runtime/JSArrayBufferView.h:
* Source/JavaScriptCore/runtime/JSAsyncDisposableStack.h:
* Source/JavaScriptCore/runtime/JSAsyncFromSyncIterator.h:
* Source/JavaScriptCore/runtime/JSCallee.h:
* Source/JavaScriptCore/runtime/JSDisposableStack.h:
* Source/JavaScriptCore/runtime/JSFunction.h:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayView.h:
* Source/JavaScriptCore/runtime/JSGlobalProxy.h:
* Source/JavaScriptCore/runtime/JSInternalFieldObjectImpl.h:
* Source/JavaScriptCore/runtime/JSLexicalEnvironment.h:
* Source/JavaScriptCore/runtime/JSMapIterator.h:
* Source/JavaScriptCore/runtime/JSModuleEnvironment.h:
* Source/JavaScriptCore/runtime/JSModuleNamespaceObject.h:
* Source/JavaScriptCore/runtime/JSModuleRecord.h:
* Source/JavaScriptCore/runtime/JSNativeStdFunction.h:
* Source/JavaScriptCore/runtime/JSRegExpStringIterator.h:
* Source/JavaScriptCore/runtime/JSRemoteFunction.h:
* Source/JavaScriptCore/runtime/JSSetIterator.h:
* Source/JavaScriptCore/runtime/JSSymbolTableObject.h:
* Source/JavaScriptCore/runtime/JSWrapForValidIterator.h:
* Source/JavaScriptCore/runtime/JSWrapperObject.h:
* Source/JavaScriptCore/runtime/ModuleProgramExecutable.h:
* Source/JavaScriptCore/runtime/ProgramExecutable.h:
* Source/JavaScriptCore/runtime/ProxyObject.h:
* Source/JavaScriptCore/runtime/ShadowRealmObject.h:
* Source/JavaScriptCore/runtime/SyntheticModuleRecord.h:
* Source/JavaScriptCore/tools/JSDollarVM.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyGlobal.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyInstance.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyMemory.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyModule.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyTable.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunction.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunctionBase.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyGCObjectBase.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyModuleRecord.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyWrapperFunction.h:
* Source/WebCore/bindings/js/JSDOMBuiltinConstructorBase.h:
* Source/WebCore/bindings/js/JSShadowRealmGlobalScopeBase.h:
* Source/WebCore/bindings/js/JSWorkerGlobalScopeBase.h:
* Source/WebCore/bindings/js/JSWorkletGlobalScopeBase.h:
Move DECLARE_VISIT_CHILDREN macro to the public section.

Canonical link: <a href="https://commits.webkit.org/298066@main">https://commits.webkit.org/298066@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f6e908535c2fb7d433de4d0770f1e8d75b798523

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114074 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33826 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24284 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120242 "") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/64843 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/115963 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34454 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42386 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/120242 "") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/41696 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117022 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27431 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102456 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/120242 "") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26616 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20585 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63946 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/106500 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96798 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20702 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123469 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/112637 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41105 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30627 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95533 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41483 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98666 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95316 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40450 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18261 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/37257 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18292 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40981 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46484 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/136837 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40606 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36641 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43906 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42359 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/wasm/js/WebAssemblyFunction.h:
```diff
@@ -105,6 +106,8 @@ class WebAssemblyFunction final : public WebAssemblyFunctionBase {
     CodePtr<JSEntryPtrTag> m_jsToWasmICJITCode;
 #endif
+
+    friend JSC::LLIntOffsetsExtractor;
```
In most of cases, we define it at the head of the class / struct definition. Can you move it?

kljohann replied:
Done, thanks for the review @Constellation.

Constellation commented on Source/JavaScriptCore/wasm/js/WebAssemblyFunctionBase.h:
```diff
@@ -76,6 +77,8 @@ class WebAssemblyFunctionBase : public JSFunction {
     Wasm::WasmOrJSImportableFunctionCallLinkInfo* m_callLinkInfo;
     WriteBarrier<JSWebAssemblyInstance> m_instance;
+
+    friend JSC::LLIntOffsetsExtractor;
```
Ditto.

kljohann committed:
[JSC] Fix wrong member access due to DECLARE_VISIT_CHILDREN macro
https://bugs.webkit.org/show_bug.cgi?id=296688

Reviewed by Yusuke Suzuki.

Using the DECLARE_VISIT_CHILDREN macro in private/protected sections can leak
subsequent members, as the macro ends with a public access specifier.

This change moves all uses of the macro into the public section of their
respective classes and selectively makes members used in existing
code public.

* Source/JavaScriptCore/runtime/JSBoundFunction.h:
Move DECLARE_VISIT_CHILDREN macro to the public section.
(JSC::JSBoundFunction::isTainted): New public getter for m_isTainted.
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::executeCall): Use new isTainted() getter.

* Source/JavaScriptCore/runtime/IntlDateTimeFormat.h:
Move DECLARE_VISIT_CHILDREN macro to the public section.
* Source/JavaScriptCore/runtime/IntlDateTimeFormatPrototype.cpp:
(JSC::handleDateTimeValue): Made TU-local as it is only used by host
functions within this file.
(JSC::IntlDateTimeFormat::handleDateTimeValue): Deleted.

* Source/JavaScriptCore/runtime/IntlNumberFormat.h:
Move DECLARE_VISIT_CHILDREN macro to the public section.
Add missing friend declarations.
(JSC::IntlNumberFormat::trailingZeroDisplayString): Made public for
use in IntlPluralRules::resolvedOptions.

* Source/JavaScriptCore/runtime/IntlPluralRules.h:
Move DECLARE_VISIT_CHILDREN macro to the public section.
Add missing friend declaration.

* Source/JavaScriptCore/runtime/RegExpObject.h:
Move DECLARE_VISIT_CHILDREN macro to the public section.
(JSC::RegExpObject::lastIndexIsWritable): Make public for
use in regExpProtoFuncSearch.

* Source/JavaScriptCore/tools/JSDollarVM.cpp:
(JSC::JSDollarVM::WasmStreamingCompiler): Add missing private
access specifier.

* Source/JavaScriptCore/API/JSCallbackObject.h:
* Source/JavaScriptCore/bytecode/UnlinkedCodeBlock.h:
* Source/JavaScriptCore/runtime/AbstractModuleRecord.h:
* Source/JavaScriptCore/runtime/EvalExecutable.h:
* Source/JavaScriptCore/runtime/GenericArgumentsImpl.h:
* Source/JavaScriptCore/runtime/IntlCollator.h:
* Source/JavaScriptCore/runtime/IntlLocale.h:
* Source/JavaScriptCore/runtime/IntlRelativeTimeFormat.h:
* Source/JavaScriptCore/runtime/JSArrayBufferView.h:
* Source/JavaScriptCore/runtime/JSAsyncDisposableStack.h:
* Source/JavaScriptCore/runtime/JSAsyncFromSyncIterator.h:
* Source/JavaScriptCore/runtime/JSCallee.h:
* Source/JavaScriptCore/runtime/JSDisposableStack.h:
* Source/JavaScriptCore/runtime/JSFunction.h:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayView.h:
* Source/JavaScriptCore/runtime/JSGlobalProxy.h:
* Source/JavaScriptCore/runtime/JSInternalFieldObjectImpl.h:
* Source/JavaScriptCore/runtime/JSLexicalEnvironment.h:
* Source/JavaScriptCore/runtime/JSMapIterator.h:
* Source/JavaScriptCore/runtime/JSModuleEnvironment.h:
* Source/JavaScriptCore/runtime/JSModuleNamespaceObject.h:
* Source/JavaScriptCore/runtime/JSModuleRecord.h:
* Source/JavaScriptCore/runtime/JSNativeStdFunction.h:
* Source/JavaScriptCore/runtime/JSRegExpStringIterator.h:
* Source/JavaScriptCore/runtime/JSRemoteFunction.h:
* Source/JavaScriptCore/runtime/JSSetIterator.h:
* Source/JavaScriptCore/runtime/JSSymbolTableObject.h:
* Source/JavaScriptCore/runtime/JSWrapForValidIterator.h:
* Source/JavaScriptCore/runtime/JSWrapperObject.h:
* Source/JavaScriptCore/runtime/ModuleProgramExecutable.h:
* Source/JavaScriptCore/runtime/ProgramExecutable.h:
* Source/JavaScriptCore/runtime/ProxyObject.h:
* Source/JavaScriptCore/runtime/ShadowRealmObject.h:
* Source/JavaScriptCore/runtime/SyntheticModuleRecord.h:
* Source/JavaScriptCore/tools/JSDollarVM.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyGlobal.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyInstance.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyMemory.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyModule.h:
* Source/JavaScriptCore/wasm/js/JSWebAssemblyTable.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunction.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyFunctionBase.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyGCObjectBase.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyModuleRecord.h:
* Source/JavaScriptCore/wasm/js/WebAssemblyWrapperFunction.h:
* Source/WebCore/bindings/js/JSDOMBuiltinConstructorBase.h:
* Source/WebCore/bindings/js/JSShadowRealmGlobalScopeBase.h:
* Source/WebCore/bindings/js/JSWorkerGlobalScopeBase.h:
* Source/WebCore/bindings/js/JSWorkletGlobalScopeBase.h:
Move DECLARE_VISIT_CHILDREN macro to the public section.

Canonical link: https://commits.webkit.org/298066@main
